security/pkg/pki/ca/selfsignedcarootcertrotator.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ca
+
+import (
+	"bytes"
+	"encoding/base64"
+	"time"
+
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"istio.io/istio/security/pkg/k8s/configmap"
+	"istio.io/istio/security/pkg/k8s/controller"
+	"istio.io/istio/security/pkg/pki/util"
+	certutil "istio.io/istio/security/pkg/util"
+	"istio.io/pkg/log"
+
+	v1 "k8s.io/api/core/v1"
+)
+
+var rootCertRotatorLog = log.RegisterScope("rootCertRotator", "Self-signed CA root cert rotator log", 0)
+
+// SelfSignedCARootCertRotator automatically checks self-signed signing root
+// certificate and rotates root certificate if it is going to expire.
+type SelfSignedCARootCertRotator struct {
+	checkInterval       time.Duration
+	caCertTTL           time.Duration
+	retryInterval       time.Duration
+	certInspector       certutil.CertUtil
+	configMapController *configmap.Controller
+	caSecretController  *controller.CaSecretController
+	caStorageNamespace  string
+	dualUse             bool
+	readSigningCertOnly bool
+	org                 string
+	rootCertFile        string
+	ca                  *IstioCA
+}
+
+type SelfSignedCARootCertRotatorConfig struct {
+	CheckInterval       time.Duration
+	caCertTTL           time.Duration
+	retryInterval       time.Duration
+	certInspector       certutil.CertUtil
+	caStorageNamespace  string
+	dualUse             bool
+	readSigningCertOnly bool
+	org                 string
+	rootCertFile        string
+	client              corev1.CoreV1Interface
+}
+
+// NewSelfSignedCARootCertRotator returns a new root cert rotator instance that
+// rotates self-signed root cert periodically.
+func NewSelfSignedCARootCertRotator(config *SelfSignedCARootCertRotatorConfig,
```

Comment:
Can you just directly pass the parameters in stead of the SelfSignedCARootCertRotatorConfig?

Reply:
Any reason can't directly pass the parameters?